src/scripts/clipperUI/panels/ratingsPanel.tsx

```diff
@@ -0,0 +1,211 @@
+				return;
+		}
+	}
+
+	/**
+	 * Get appropriate dialog panel buttons for the panel (with its internal states) provided
+	 */
+	private getDialogButtons(panel: RatingsPanelClass): DialogButton[] {
+		let stage: RatingsPromptStage = panel.state.userSelectedRatingsPromptStage;
+		let clipperState: ClipperState = panel.props.clipperState;
+		let clientType: ClientType = clipperState.clientInfo.clipperType;
+
+		let buttons: DialogButton[] = [];
+
+		switch (stage) {
+			case RatingsPromptStage.Init:
+				buttons.push({
+					id: Constants.Ids.ratingsButtonInitYes,
+					label: Localization.getLocalizedString("WebClipper.Label.Ratings.Button.Init.Positive"),
+					handler: () => {
+						RatingsHelper.setDoNotPromptStatus();
+
+						let rateUrl: string = RatingsHelper.getRateUrlIfExists(clientType);
+						if (rateUrl) {
+							panel.setState({
+								userSelectedRatingsPromptStage: RatingsPromptStage.Rate
+							});
+						} else {
+							panel.setState({
+								userSelectedRatingsPromptStage: RatingsPromptStage.End
+							});
+						}
+					}
+				}, {
+						id: Constants.Ids.ratingsButtonInitNo,
+						label: Localization.getLocalizedString("WebClipper.Label.Ratings.Button.Init.Negative"),
+						handler: () => {
+							let lastSeenVersion: string = Clipper.Storage.getCachedValue(Constants.StorageKeys.lastSeenVersion);
+							let badRatingAlreadyOccurred: boolean = RatingsHelper.setLastBadRating(Date.now().toString(), lastSeenVersion);
+							if (badRatingAlreadyOccurred) {
+								// setting this to prevent additional ratings prompts after the second bad rating
+								RatingsHelper.setDoNotPromptStatus();
+							}
+							let feedbackUrl: string = RatingsHelper.getFeedbackUrlIfExists(clipperState);
+							if (feedbackUrl) {
+								panel.setState({
+									userSelectedRatingsPromptStage: RatingsPromptStage.Feedback
+								});
+							} else {
+								panel.setState({
+									userSelectedRatingsPromptStage: RatingsPromptStage.End
+								});
+							}
+						}
+					});
+				break;
+			case RatingsPromptStage.Rate:
+				let rateUrl: string = RatingsHelper.getRateUrlIfExists(clientType);
+				if (rateUrl) {
+					buttons.push({
+						id: Constants.Ids.ratingsButtonRateYes,
+						label: Localization.getLocalizedString("WebClipper.Label.Ratings.Button.Rate"),
+						handler: () => {
+							window.open(rateUrl, "_blank");
+
+							panel.setState({
+								userSelectedRatingsPromptStage: RatingsPromptStage.End
+							});
+						}
+					}, {
+						id: Constants.Ids.ratingsButtonRateNo,
+						label: Localization.getLocalizedString("WebClipper.Label.Ratings.Button.NoThanks"),
+						handler: () => {
+							panel.setState({
+								userSelectedRatingsPromptStage: RatingsPromptStage.None
+							});
+						}
+					});
+				} else {
+					// this shouldn't happen
+					panel.setState({
+						userSelectedRatingsPromptStage: RatingsPromptStage.None
+					});
+				}
+				break;
+			case RatingsPromptStage.Feedback:
```

Comment:
Can you split these into functions for each state for readability? The logic looks good though.